Agent/Utilities/NRAutoLogCollector.m

```diff
@@ -0,0 +1,168 @@
+//
+//  NRAutoLogCollector.m
+//  Agent
+//
+//  Created by Mike Bruin on 10/9/24.
+//  Copyright © 2024 New Relic. All rights reserved.
+//
+
+#import "NRAutoLogCollector.h"
+#import "NRLogger.h"
+
+int saved_stdout;
+int saved_stderr;
+int stdoutPipe[2];
+int stderrPipe[2];
+static BOOL hasRedirectedStdOut = false;
+
+@interface NRAutoLogCollector()
+
+@end
+
+@implementation NRAutoLogCollector
+
++ (BOOL) redirectStandardOutputAndError {
+    if (hasRedirectedStdOut){
+        return true;
+    }
+    // Create pipes for stdout and stderr
+    if (pipe(stdoutPipe) == -1 || pipe(stderrPipe) == -1) {
+        return false;
+    }
+
+    // Save the original stdout and stderr file descriptors
+    saved_stdout = dup(fileno(stdout));
+    saved_stderr = dup(fileno(stderr));
+    if (saved_stdout == -1 || saved_stderr == -1) {
+        close(stdoutPipe[0]);
+        close(stdoutPipe[1]);
+        close(stderrPipe[0]);
+        close(stderrPipe[1]);
+        return false;
+    }
+
+    // Redirect stdout and stderr to the write ends of the pipes
+    if (dup2(stdoutPipe[1], fileno(stdout)) == -1 || dup2(stderrPipe[1], fileno(stderr)) == -1) {
+        close(stdoutPipe[0]);
+        close(stdoutPipe[1]);
+        close(stderrPipe[0]);
+        close(stderrPipe[1]);
+        close(saved_stdout);
+        close(saved_stderr);
+        return false;
+    }
+    close(stdoutPipe[1]); // Close the original write end of the stdout pipe
+    close(stderrPipe[1]); // Close the original write end of the stderr pipe
+
```

Comment:
You should investigate disabling buffering of the pipes

Reply:
Yes, I believe that's correct.

Reply:
I tried adding that but it seems to mess with the accuracy of the logs. Giving more logs out of order and having partial logs.